asapdiscovery-ml/asapdiscovery/ml/dataset.py

```diff
@@ -547,22 +564,35 @@ def __init__(
             task_names=[],
         )
 
-        self.compounds = {}
-        self.structures = []
-        for i, (compound_id, g) in enumerate(zip(all_compound_ids, dataset)):
-            # Need a tuple to match DockedDataset, but the graph objects aren't
-            #  attached to a protein structure at all
-            compound = ("NA", compound_id)
+        for compound_id, g in zip(all_compound_ids, dataset):
+            self.compounds_dict[compound_id] = g[1]
+            self.graphs.append(g[1])
+            self.smiles_dict[g[0]] = g[1]
 
-            # Add data
-            try:
-                self.compounds[compound].append(i)
-            except KeyError:
-                self.compounds[compound] = [i]
-            self.structures.append(
-                {
-                    "smiles": g[0],
-                    "g": g[1],
-                    "compound": compound,
-                }
+    def __len__(self):
+        return len(self.graphs)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, str):
+            return self.smiles_dict[idx]
+        elif isinstance(idx, int):
+            return self.graphs[idx]
+        elif isinstance(idx, list):
+            return [self.graphs[i] for i in idx]
+        elif isinstance(idx, slice):
+            # Use the slice indices to slice the list of graphs and create a new GraphInferenceDataset
+            start, stop, step = idx.indices(len(self.graphs))
+            indices = range(start, stop, step)
+            new_dataset = GraphInferenceDataset(
```

Comment:
I'm not sure which is better here, but `pytorch` has a [`Subset` class](https://pytorch.org/docs/stable/data.html#torch.utils.data.Subset) that I think does what we'd want in the case of getting a `slice` here, and it also prevents any data copying I believe

Reply:
Using subset now, good tip seems very handy.